Pcap++/header/PcapLiveDeviceList.h

```diff
@@ -23,20 +24,26 @@ namespace pcpp
 	class PcapLiveDeviceList
 	{
 	private:
-		std::vector<PcapLiveDevice*> m_LiveDeviceList;
+		std::vector<std::shared_ptr<PcapLiveDevice>> m_LiveDeviceList;
```

Comment:
>This implementation uses shared pointers to store the live devices in preparation for point 4 of https://github.com/seladb/PcapPlusPlus/issues/1431.

From the PR description. I plan on having the API return shared pointers.

Reply:
I'm not sure I fully understand point 4, but I'm not sure there's a good reason to unify the APIs of `PcapLiveDeviceList` and `PcapRemoteDeviceList`, from multiple reasons:
- `PcapLiveDeviceList` is a singleton because it finds all the interfaces on the machine. However `PcapRemoteDeviceList` is not a singleton because it encapsulates remote devices on other machines so a user can have multiple instances of it. The API is quite different so not sure it's a good idea to merge it
- Remote devices are a niche that isn't used a lot, whereas pcap devices is probably the most popular feature of PcapPlusPlus. I wouldn't mix between them

Reply:
Well, one of the reasons I started thing about it is that both `PcapLiveDEviceList` and `PcapRemoteDeviceList` share a couple similarities and have some code duplication.
- Both store their devices in `vector<DevicePtr>`
- The methods `getPcapLiveDeviceByIp` and `getPcapRemoteDeviceByIP` have practically duplicate implementation.

One of the reasons for unifying the API is having an internal base class `template<DeviceType> PcapLiveDeviceListBase` (name subject to change) that implements the `getPcapDeviceByIP` methods that currently have duplicate implementation.

The current methods could still be kept and just call the new functions too.

Reply:
Also, even if the API is not unified, I still have the proposal to get the API to return `shared_ptr` instead of raw pointers so you don't suddenly get invalid pointers. (refresh, or in case of `RemoteDeviceList` just the list going out of scope).

Reply:
to be honest, I wouldn't spend time on it because although there is some code duplication, it's not a lot and we can keep the code simple. If you really insist we can discuss it more...

Reply:
Hmm, ok then. Gonna shelve point 4 for now.

Reply:
> PcapLiveDeviceList is a singleton

@seladb Btw, if the class is a singleton, why does it have a `clone` method?

Reply:
Changed to `unique_ptr`. ad77379

Reply:
> > PcapLiveDeviceList is a singleton
> 
> @seladb Btw, if the class is a singleton, why does it have a `clone` method?

I agree this API is a bit weird 🙈 
The reason is that users wanted to capture traffic from the same interface more than once, here is the GitHub issue: https://github.com/seladb/PcapPlusPlus/pull/766

The idea was to allow cloning a live device, or to allow cloning all devices...